zero_bin/rpc/src/native/txn.rs

```diff
@@ -357,3 +398,228 @@ fn prestate_tracing_options(diff_mode: bool) -> GethDebugTracingOptions {
     }
 }
+
+/// Tracing options for the `debug_traceTransaction` call to get structlog.
+/// Used for filling JUMPDEST table.
+fn structlog_tracing_options() -> GethDebugTracingOptions {
+    GethDebugTracingOptions {
+        config: GethDefaultTracingOptions {
+            disable_stack: Some(false),
+            disable_memory: Some(true),
+            disable_storage: Some(true),
+            ..GethDefaultTracingOptions::default()
+        },
+        tracer: None,
+        ..GethDebugTracingOptions::default()
+    }
+}
+
+/// Generate at JUMPDEST table by simulating the call stack in EVM,
+/// using a Geth structlog as input.
+async fn generate_jumpdest_table(
+    tx: &Transaction,
+    structlog_trace: &DefaultFrame,
+    tx_traces: &HashMap<Address, TxnTrace>,
+) -> anyhow::Result<JumpDestTableWitness> {
+    info!("Generating JUMPDEST table for tx: {}", tx.hash);
+    let mut jumpdest_table = JumpDestTableWitness::default();
+
+    if structlog_trace.struct_logs.is_empty() {
+        return Ok(jumpdest_table);
+    };
+
+    let callee_addr_to_code_hash: HashMap<Address, H256> = tx_traces
+        .iter()
+        .map(|(callee_addr, trace)| (callee_addr, &trace.code_usage))
+        .filter(|(_callee_addr, code_usage)| code_usage.is_some())
+        .map(|(callee_addr, code_usage)| {
+            (*callee_addr, code_usage.as_ref().unwrap().get_code_hash())
+        })
+        .collect();
+
+    let to_address: Address = tx
+        .to
+        .unwrap_or_else(|| panic!("No `to`-address for tx: {}.", tx.hash));
+
+    // Guard against transactions to a non-contract address.
+    if callee_addr_to_code_hash.contains_key(&to_address).not() {
+        return Ok(jumpdest_table);
+    }
+    let entrypoint_code_hash: H256 = callee_addr_to_code_hash[&to_address];
+
+    // `None` encodes that previous `entry`` was not a JUMP or JUMPI with true
+    // condition, `Some(jump_target)` encodes we came from a JUMP or JUMPI with
+    // true condition and target `jump_target`.
+    let mut prev_jump = None;
+
+    // Contains the previous op.
+    let mut prev_op = "";
+
+    // Call depth of the previous `entry`.  We initialize to 0 as this compares
+    // smaller to 1.
+    let mut prev_depth = 0;
+    // The next available context. Starts at 1. Never decrements.
+    let mut next_ctx_available = 1;
+    // Immediately use context 1;
+    let mut call_stack = vec![(entrypoint_code_hash, next_ctx_available)];
+    next_ctx_available += 1;
+
+    for (step, entry) in structlog_trace.struct_logs.iter().enumerate() {
+        let op = entry.op.as_str();
+        let curr_depth = entry.depth;
+
+        let exception_occurred = prev_entry_caused_exception(prev_op, prev_depth, curr_depth);
+        if exception_occurred {
+            call_stack.pop().expect("Call stack was empty at POP.");
+        }
+
+        debug_assert!(entry.depth as usize <= next_ctx_available);
+        let (code_hash, ctx) = call_stack.last().unwrap();
```

Comment:
This can panic

Reply:
Yeah so in general I should never panic.

Reply:
Replaced by `ensure!`.